File: 백준/Silver/9012. 괄호/괄호.py
n = int(input())

for i in range(n):
    count = 0 # (를 count
    vps = input()
    for j in vps: # vps안에 (가 있으면 +1 아니면 -1
        if j == '(':
            count += 1
        else:
            count -= 1
            if count < 0: # count가 음수로 가는 경우에 vps를 만족하지 못함
                break
    if count == 0: # count가 0이 아니면 vps만족하지 못하므로
        print('YES')
    else:
        print('NO')

File: 백준/Silver/9012. 괄호/괄호.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine());
        
        for (int i = 0; i < T; i++) {
            String input = br.readLine();
            Deque<Character> deque = new ArrayDeque<>();
            boolean invalid = true;
            
            for (int j = 0; j < input.length(); j++) {
                if (input.charAt(j) == '(') {
                    deque.addLast('(');
                } else {
                    if (!deque.isEmpty()) {
                        deque.pollLast();
                    } else {
                        System.out.println("NO");
                        invalid = false;
                        break;
                    }
                }
            }
            
            if (invalid) {
                if (deque.isEmpty()) {
                    System.out.println("YES");
                } else {
                    System.out.println("NO");
                }
            }
        }
    }
}